### PR TITLE
added LightProfiniteInjective

### PR DESCRIPTION
**Pull request by LennyTaelman**
Added file LightProfiniteInjective proving that nonempty light profinite spaces are injective

### CODE REVIEW

File: LeanCondensed/Projects/LightProfiniteInjective.lean [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+lemma light_to_final_split_epi (X : LightProfinite.{u}) [Nonempty X] :
+    IsSplitEpi (Limits.terminalIsTerminal.from X) := by
+  let s : ⊤_ LightProfinite ⟶ X :=
+    { toFun := fun _ => Nonempty.some inferInstance, continuous_toFun := continuous_const}
+  exact IsSplitEpi.mk' { section_ := s, id := Limits.terminal.hom_ext _ _ }
+
+lemma light_to_final_epi (X : LightProfinite.{u}) [Nonempty X] :
+    Epi (Limits.terminalIsTerminal.from X) := by
+  haveI := light_to_final_split_epi X
+  exact IsSplitEpi.epi (Limits.terminalIsTerminal.from X)
```
Comment (dagurtomas) [inline LeanCondensed/Projects/LightProfiniteInjective.lean]:
```suggestion
instance {P : TopCat.{u} → Prop} [CompHausLike.HasProp P PUnit.{u+1}]
    (X : CompHausLike.{u} P) [Nonempty X] :
    IsSplitEpi (CompHausLike.isTerminalPUnit.from X) := IsSplitEpi.mk'
  { section_ := CompHausLike.const _ (Nonempty.some inferInstance)
    id := CompHausLike.isTerminalPUnit.hom_ext _ _ }
```

File: LeanCondensed/Projects/LightProfiniteInjective.lean [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+
+lemma profinite_injective_of_finite (S : Profinite.{u}) [Nonempty S] [Finite S]:
+    Injective (S) := by
+  constructor
+  intro X Y g f f_mono
+  let f' := Limits.terminalIsTerminal.from S
+  haveI : Epi  f' := to_final_epi S
+  let g' := Limits.terminalIsTerminal.from Y
+  obtain ⟨k, _, h2⟩ := profinite_key_lifting_lemma _ _ S _ f f' g g' (Limits.terminal.hom_ext _ _)
+  exact ⟨k, h2⟩
```
Comment (dagurtomas) [inline LeanCondensed/Projects/LightProfiniteInjective.lean]:
```suggestion
instance light_injective_of_finite (S : LightProfinite.{u}) [Nonempty S] [Finite S] : 
    Injective (S) where
  factors {X Y} g f _ := by
    let f' := CompHausLike.isTerminalPUnit.from S
    let g' := CompHausLike.isTerminalPUnit.from Y
    obtain ⟨k, _, h2⟩ := light_key_lifting_lemma _ _ S _ f f' g g'
      (CompHausLike.isTerminalPUnit.hom_ext _ _)
    exact ⟨k, h2⟩

instance profinite_injective_of_finite (S : Profinite.{u}) [Nonempty S] [Finite S] :
    Injective (S) where
  factors {X Y} g f _ := by
    let f' := CompHausLike.isTerminalPUnit.from S
    let g' := CompHausLike.isTerminalPUnit.from Y
    obtain ⟨k, _, h2⟩ := profinite_key_lifting_lemma _ _ S _ f f' g g'
      (CompHausLike.isTerminalPUnit.hom_ext _ _)
    exact ⟨k, h2⟩
```

File: LeanCondensed/Projects/LightProfiniteInjective.lean [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+    congr
+    ext n
+    dsimp only [Cone.extensions_app, NatTrans.comp_app, Functor.const_map_app,
+      NatTrans.ofOpSequence_app]
+    rw [h_down]
+  rw [← hlim]
+  apply S.asLimit.uniq (k_cone.extend f)
+  intro n
+  simp only [Category.assoc, IsLimit.fac, Cone.extend_π,  Cone.extensions_app,
+    NatTrans.comp_app,  Functor.const_map_app]
```
Comment (dagurtomas) [inline LeanCondensed/Projects/LightProfiniteInjective.lean]:
```suggestion
instance injective_of_light (S : LightProfinite.{u}) [Nonempty S]: Injective S where
  factors {X Y} g f h := by
    -- help the instance inference a bit
    haveI (n : ℕ) : Finite (S.component n) :=
      inferInstanceAs (Finite (FintypeCat.toLightProfinite.obj _))
    haveI : Nonempty (S.component 0) := Nonempty.map (S.proj 0).toFun inferInstance
    haveI (n : ℕ) : Epi (S.transitionMap n) := (LightProfinite.epi_iff_surjective _).mpr
      (S.surjective_transitionMap n)
    -- base step of the induction: find k0 : Y ⟶ S.component 0
    obtain ⟨k0, h_down0⟩ :=  Injective.factors (g ≫ S.proj 0) f
    have h_comm0 : f ≫ k0 = g ≫ S.proj 1 ≫ S.transitionMap 0 := by
      rw [h_down0]
      exact congrArg _ (S.proj_comp_transitionMap 0).symm
    -- key part of induction step: next produces k n+1 out of k n, see diagram above
    have h_step (n : ℕ) (k : Y ⟶ S.component n) (h_down : f ≫ k = g ≫ S.proj n) :
      ∃ k' : Y ⟶ S.component (n+1), k' ≫ S.transitionMap n = k ∧ f ≫ k' = g ≫ S.proj (n+1) := by
      have h_comm : f ≫ k = g ≫ S.proj (n+1) ≫ S.transitionMap n := by
        rw [h_down]
        exact congrArg _ (S.proj_comp_transitionMap n).symm
      exact light_key_lifting_lemma _ _ _ _ f (S.transitionMap n) (g ≫ (S.proj (n+1))) k h_comm
    let lifts (n : ℕ) := { k : Y ⟶ S.component n // f ≫ k = g ≫ S.proj n }
    let next (n : ℕ) : lifts n → lifts (n+1) :=
      fun k ↦ ⟨Classical.choose (h_step n k.val k.property),
        (Classical.choose_spec (h_step n k.val k.property)).2⟩
    -- now define sequence of lifts using induction
    let k_seq (n : Nat) : lifts n := Nat.rec ⟨k0, h_down0⟩ next n
    -- h_up and h_down are the required commutativity properties
    have h_down (n : ℕ) : f ≫ (k_seq n).val = g ≫ S.proj n := (k_seq n).property
    have h_up (n : ℕ) : (k_seq (n+1)).val ≫ S.transitionMap n = (k_seq n).val := by
      induction' n with n ih
      · exact (Classical.choose_spec (h_step 0 k0 h_down0)).1
      · exact (Classical.choose_spec (h_step (n+1) (k_seq (n+1)).val (k_seq (n+1)).property)).1
    let k_cone : Cone S.diagram :=
      { pt := Y, π := NatTrans.ofOpSequence (fun n ↦ (k_seq n).val) (fun n ↦ (h_up n).symm) }
    -- now the induced map Y ⟶ S = lim S.component is the desired map
    use S.asLimit.lift k_cone
    let g_cone : Cone S.diagram :=
      { pt := X, π := NatTrans.ofOpSequence (fun n ↦ g ≫ S.proj n) (fun n ↦ by
        simp only [Functor.const_obj_obj, Functor.const_obj_map, Category.id_comp]
        exact congrArg _ (S.proj_comp_transitionMap n).symm) }
    have hg : g = S.asLimit.lift g_cone := by
      apply S.asLimit.uniq g_cone
      intro n
      simp only [NatTrans.ofOpSequence_app]
    rw [hg]
    have hlim : S.asLimit.lift (k_cone.extend f) = S.asLimit.lift g_cone := by
      unfold Cone.extend
      congr
      ext n
      dsimp only [Cone.extensions_app, NatTrans.comp_app, Functor.const_map_app,
        NatTrans.ofOpSequence_app]
      rw [h_down]
    rw [← hlim]
    apply S.asLimit.uniq (k_cone.extend f)
    intro n
    simp only [Category.assoc, IsLimit.fac, Cone.extend_π,  Cone.extensions_app,
      NatTrans.comp_app,  Functor.const_map_app]
```

and remove the instance below

**Comment by dagurtomas** [comment]
Could you go through the whole file and squeeze all the non-terminal simps (this means replace every `simp` in the middle of a proof with the `simp only` given by `simp?`).

Also, please put everything in the `LightProfinite` namespace. We can polish this a lot more before PR-ing to mathlib, but it's good enough for now. Thanks!

**Comment by LennyTaelman** [comment]
I think I've made all the changes. I've also renamed the lifting lemmas in to extension lemmas, which I think is more accurate and intuitive. 

Everything is now in one LightProfinite namespace, but eventually I guess the purely profinite stuff (mostly the first half) has to go in Profinite, and the 'nonempty to final is epi' in CompHausLike?

**Pull request merged**